tests/run/lazy-impl.scala

```diff
@@ -0,0 +1,220 @@
+ *
+ *   - no slot other than the field itself is needed
+ *   - no locks are shared among lazy val initializations and between lazy val initializations
+ *     and normal code
+ *   - no deadlocks (other than those inherent in user code)
+ *   - synchronized code is executed only if there is contention
+ *   - simpler than current Dotty scheme
+ *
+ *  Disadvantages:
+ *
+ *   - does not work for local lazy vals (but maybe these could be unsynchronized anyway?)
+ *   - lazy vals of primitive types are boxed
+ */
+import sun.misc.Unsafe._
+
+class C {
+  def init(name: String) = {
+    Thread.sleep(10)
+    println(s"initialize $name"); "result"
+  }
+
+  private[this] var _x: AnyRef = null
+
+  // Expansion of:  lazy val x: String = init
+
+  def x: String = {
+    val current = _x
+    if (current.isInstanceOf[String])
+      current.asInstanceOf[String]
+    else
+      x$lzy_compute
+  }
+
+  def x$lzy_compute: String = {
+    val current = _x
+    if (current.isInstanceOf[String])
+      current.asInstanceOf[String]
+    else {
+      val offset = C.x_offset
+      if (current == null) {
+        if (LazyRuntime.isUnitialized(this, offset))
+          LazyRuntime.initialize(this, offset, init("x"))
+      }
+      else
+        LazyRuntime.awaitInitialized(this, offset, current)
+      x$lzy_compute
+    }
+  }
+
+  // Compare with bytecodes for regular lazy val:
+  lazy val y = init("y")
+}
+
+object C {
+  import LazyRuntime.fieldOffset
+  val x_offset = fieldOffset(classOf[C], "_x")
+}
+
+object LazyRuntime {
+  val Evaluating = new LazyControl()
+
+  private val unsafe: sun.misc.Unsafe = {
+    val f: java.lang.reflect.Field = classOf[sun.misc.Unsafe].getDeclaredField("theUnsafe");
+    f.setAccessible(true)
+    f.get(null).asInstanceOf[sun.misc.Unsafe]
+  }
+
+  def fieldOffset(cls: Class[_], name: String): Long = {
+    val fld = cls.getDeclaredField(name)
+    fld.setAccessible(true)
+    unsafe.objectFieldOffset(fld)
+  }
+
+  def isUnitialized(base: Object, offset: Long): Boolean =
+    unsafe.compareAndSwapObject(base, offset, null, Evaluating)
+
+  def initialize(base: Object, offset: Long, result: Object): Unit =
+    if (!unsafe.compareAndSwapObject(base, offset, Evaluating, result)) {
+      val lock = unsafe.getObject(base, offset).asInstanceOf[Waiting]
+      unsafe.putObject(base, offset, result)
+      lock.release(result)
+    }
+
+  def awaitInitialized(base: Object, offset: Long, current: Object): Unit =
+    if (current.isInstanceOf[Waiting])
+      unsafe.putObject(base, offset, current.asInstanceOf[Waiting].awaitRelease())
+    else
+      unsafe.compareAndSwapObject(base, offset, Evaluating, new Waiting)
+}
+
+class LazyControl
+
+class Waiting extends LazyControl {
+
+  private var done = false
+  private var result: AnyRef = _
+
+  def release(result: AnyRef) = synchronized {
+    this.result = result
+    done = true
+    notifyAll()
+  }
+
+  def awaitRelease(): AnyRef = synchronized {
+    if (!done) wait()
```

Comment:
To handle spurious wake-ups:
```suggestion
    while (!done) wait()
```

Reply:
I don't see how a spurious wakeup could happen?

Reply:
That's what makes them spurious :)

Reply:
Are you referring to something like this?

> https://softwareengineering.stackexchange.com/questions/186842/spurious-wakeups-explanation-sounds-like-a-bug-that-just-isnt-worth-fixing-is

That something like this could happen was news to me! Yes, in that world we need a `while`.

Reply:
@odersky yes, exactly that.

Reply:
You also need to consider what would/should happen if someone fires a `Thread.interrupt` at any threads waiting on initialisation.